File: .cursor/skills/sleuths/sleuth/src/discover.rs
use crate::config::{cursor_projects_dir, SleuthsConfig};
use crate::slug::slug_from_path;
use anyhow::{Context, Result};
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::SystemTime;

#[derive(Debug, Clone)]
pub struct TranscriptSegment {
    pub transcript_id: String,
    pub relative_path: String,
    pub absolute_path: PathBuf,
    pub mtime: SystemTime,
}

pub fn resolve_slugs(project_root: &Path, config: &SleuthsConfig) -> Result<Vec<String>> {
    let mut slugs = vec![slug_from_path(project_root)];

    if project_root.join(".git").exists() {
        if let Ok(extra) = git_worktree_paths(project_root) {
            for path in extra {
                let s = slug_from_path(&path);
                if !slugs.contains(&s) {
                    slugs.push(s);
                }
            }
        }
    }

    for s in &config.transcripts.extra_transcript_slugs {
        if !slugs.contains(s) {
            slugs.push(s.clone());
        }
    }

    slugs.sort();
    slugs.dedup();
    eprintln!("Resolved transcript slugs: {}", slugs.join(", "));
    Ok(slugs)
}

fn git_worktree_paths(project_root: &Path) -> Result<Vec<PathBuf>> {
    let output = Command::new("git")
        .args(["worktree", "list", "--porcelain"])
        .current_dir(project_root)
        .output()
        .context("git worktree list")?;

    if !output.status.success() {
        anyhow::bail!(
            "git worktree list failed: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }

    let mut paths = Vec::new();
    let text = String::from_utf8_lossy(&output.stdout);
    for line in text.lines() {
        if let Some(path) = line.strip_prefix("worktree ") {
            paths.push(PathBuf::from(path));
        }
    }
    Ok(paths)
}

pub fn discover_segments(slugs: &[String]) -> Result<Vec<TranscriptSegment>> {
    let projects = cursor_projects_dir();
    let mut segments = Vec::new();

    for slug in slugs {
        let root = projects.join(slug).join("agent-transcripts");
        if !root.is_dir() {
            continue;
        }

        for entry in walkdir::WalkDir::new(&root)
            .follow_links(false)
            .into_iter()
            .filter_map(|e| e.ok())
        {
            let path = entry.path();
            if !path.is_file() {
                continue;
            }
            if path.extension().and_then(|e| e.to_str()) != Some("jsonl") {
                continue;
            }

            let transcript_id = session_id_from_path(&root, path)?;
            let relative_path = path
                .strip_prefix(&root)
                .context("strip agent-transcripts prefix")?
                .to_string_lossy()
                .replace('\\', "/");

            let mtime = std::fs::metadata(path)
                .and_then(|m| m.modified())
                .unwrap_or(SystemTime::UNIX_EPOCH);

            segments.push(TranscriptSegment {
                transcript_id,
                relative_path,
                absolute_path: path.to_path_buf(),
                mtime,
            });
        }
    }

    segments.sort_by_key(|s| s.mtime);
    Ok(segments)
}

fn session_id_from_path(agent_transcripts: &Path, jsonl: &Path) -> Result<String> {
    let rel = jsonl
        .strip_prefix(agent_transcripts)
        .context("relative to agent-transcripts")?;
    let parts: Vec<_> = rel.components().collect();
    match parts.first() {
        Some(std::path::Component::Normal(id)) => Ok(id.to_string_lossy().into_owned()),
        _ => anyhow::bail!("unexpected jsonl path: {}", jsonl.display()),
    }
}

File: .cursor/skills/sleuths/sleuth/discover.py
from __future__ import annotations

import subprocess
import sys
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path

from sleuth.config import SleuthsConfig, cursor_projects_dir
from sleuth.slug import slug_from_path


@dataclass
class TranscriptSegment:
    transcript_id: str
    relative_path: str
    absolute_path: Path
    mtime: float


def resolve_slugs(project_root: Path, config: SleuthsConfig) -> list[str]:
    slugs = [slug_from_path(project_root)]

    if (project_root / ".git").exists():
        try:
            for path in _git_worktree_paths(project_root):
                s = slug_from_path(path)
                if s not in slugs:
                    slugs.append(s)
        except (OSError, subprocess.CalledProcessError):
            pass

    for s in config.transcripts.extra_transcript_slugs:
        if s not in slugs:
            slugs.append(s)

    slugs = sorted(set(slugs))
    print(f"Resolved transcript slugs: {', '.join(slugs)}", file=sys.stderr)
    return slugs


def _git_worktree_paths(project_root: Path) -> list[Path]:
    output = subprocess.run(
        ["git", "worktree", "list", "--porcelain"],
        cwd=project_root,
        capture_output=True,
        text=True,
        check=True,
    )
    paths: list[Path] = []
    for line in output.stdout.splitlines():
        if line.startswith("worktree "):
            paths.append(Path(line[len("worktree ") :]))
    return paths


def discover_segments(slugs: list[str]) -> list[TranscriptSegment]:
    projects = cursor_projects_dir()
    segments: list[TranscriptSegment] = []

    for slug in slugs:
        root = projects / slug / "agent-transcripts"
        if not root.is_dir():
            continue

        for path in root.rglob("*.jsonl"):
            if not path.is_file():
                continue
            transcript_id = _session_id_from_path(root, path)
            relative_path = path.relative_to(root).as_posix()
            try:
                mtime = path.stat().st_mtime
            except OSError:
                mtime = 0.0
            segments.append(
                TranscriptSegment(
                    transcript_id=transcript_id,
                    relative_path=relative_path,
                    absolute_path=path,
                    mtime=mtime,
                )
            )

    segments.sort(key=lambda s: s.mtime)
    return segments


def _session_id_from_path(agent_transcripts: Path, jsonl: Path) -> str:
    rel = jsonl.relative_to(agent_transcripts)
    parts = rel.parts
    if not parts:
        raise ValueError(f"unexpected jsonl path: {jsonl}")
    return parts[0]
